### PR TITLE
feat: PYOK LLM integration, LinkedIn import, job post tailoring (v0.3.0)

### LINKED ISSUE #8: Feature: allow PYOK (pass your own key) so we can link to an LLM to assist with NLP for resume review and color scheme generation based upon provided company URL
As a user, I would like to be able to pass my own key to assist with resume review and color scheme generation. Please allow any variety of keys if possible such that we can call LLMs using the key: the main ones should be Claude, Codex, GLM/Z.AI, and any others which provide an easy API access for LLM generation

### LINKED ISSUE #9: Feature: provide a --linkedin argument such that I can parse information to create or update my resume from LinkedIn
As a user, I would like to be able to pass my LinkedIn such that I can generate or update my resume directly from my LinkedIn with the assistance of an LLM. It should take either the LinkedIn url to pull and cache the information for parsing, or a file which has the LinkedIn information already downloaded or pasted.

### LINKED ISSUE #10: Feature: provide a --job-post argument that takes either a url or a file with a job post so we can parse it and use it to review or update our resume using a linked LLM
As a user, I would like to be able to provide a job posting to uniquely update my resume for a specific job to which I am applying. This should pair with the PYOK to have LLM assist with the updates. This can also assist HR with ensuring a resume meets a the criteria of the job post, with an optional --report option to report on each resume reviewed.

### PR DESCRIPTION
**Pull request by athola**
## Summary

- **PYOK (Plug Your Own Key) LLM infrastructure** (#8): Provider-agnostic LLM integration with `LLMProvider` protocol, `LiteLLMProvider` wrapper, automatic API key resolution from environment variables, and graceful degradation via `requires_llm` decorator when `[llm]` extra is not installed.
- **LinkedIn profile import** (#9): `simple-resume import` CLI subcommand that parses HTML and plain-text LinkedIn profile exports and converts them to simple-resume YAML format.
- **Job post tailoring** (#10): `simple-resume tailor` CLI subcommand with resume-job gap analysis (`analyze_gaps`), optional LLM-assisted tailoring via `build_tailoring_prompt`, and `--report` flag for gap-only analysis without LLM.

Fixes #8, Fixes #9, Fixes #10

## Architecture

Follows Functional Core / Imperative Shell pattern:

| Layer | New Modules | Purpose |
|-------|------------|---------|
| Core | `core/llm/protocol.py` | `LLMProvider` ABC + `LLMConfig` dataclass |
| Core | `core/llm/prompts.py` | Prompt templates (pure data) |
| Core | `core/ats/tailor.py` | Gap analysis + tailoring prompt builder |
| Core | `core/importers/linkedin.py` | LinkedIn data → resume schema transform |
| Shell | `shell/llm/client.py` | `LiteLLMProvider` (litellm wrapper) |
| Shell | `shell/llm/config.py` | API key resolution (CLI, env vars) |
| Shell | `shell/llm/gate.py` | `requires_llm` decorator, availability check |
| Shell | `shell/importers/linkedin_fetcher.py` | HTML/text LinkedIn parser |
| Shell | `shell/cli/_import.py` | `import` CLI subcommand |
| Shell | `shell/cli/_tailor.py` | `tailor` CLI subcommand |

## New Optional Dependencies

```toml
[project.optional-dependencies]
llm = ["litellm>=1.0,<2.0"]
linkedin = ["beautifulsoup4>=4.12,<5.0"]
all = ["simple-resume[llm,linkedin]"]
```

## Test plan

- [x] 1970 tests passing (115 new), 94.81% coverage
- [x] All pre-commit hooks pass (ruff, mypy, bandit, ty, architecture checks)
- [x] Wheel install test passes
- [x] `simple-resume --version` reports 0.3.0

### CODE REVIEW

**Comment by athola** [comment]
## PR Review: feat: PYOK LLM integration, LinkedIn import, job post tailoring (v0.3.0)

**Branch**: `pyok-0.3.0` → `main` | **Files**: 33 changed (+4,256) | **Tests**: 1970 passing (115 new)

### Verdict: APPROVE with non-blocking suggestions

---

### Scope Validation

PR implements 3 issues (#8, #9, #10) matching the design plan in `docs/plans/2026-02-24-pyok-linkedin-jobpost-design.md`. All features are in scope. Version bump to 0.3.0 is consistent across `pyproject.toml`, `__init__.py`, and runtime.

### Architecture ✅

Clean FCIS (Functional Core / Imperative Shell) separation:
- **Core** (pure): `protocol.py`, `prompts.py`, `tailor.py`, `linkedin.py` — zero I/O
- **Shell** (side effects): `client.py`, `config.py`, `gate.py`, `linkedin_fetcher.py`, CLI handlers
- Feature gating via `requires_llm` decorator is well-designed
- Optional deps properly configured: `[llm]`, `[linkedin]`, `[all]`

### Blocking Issues

None. The PR is clean and well-structured.

### Non-Blocking Improvements

**1. Missing LLM response None check** (`shell/llm/client.py:70`)
```python
return response.choices[0].message.content  # Could be None
```
If the model returns no content, this silently returns `None` instead of `str`. Consider:
```python
content = response.choices[0].message.content
if content is None:
    raise LLMError("LLM returned empty response", provider="litellm", model=self.config.model)
return content
```

**2. Tailored output not validated** (`shell/cli/_tailor.py:159-160`)
The design doc specifies "validate tailored YAML" but the LLM output is written raw to disk. If the LLM returns malformed YAML or markdown-wrapped output, the user gets a broken file. Consider parsing with `yaml.safe_load()` before writing, or at minimum adding a comment that validation is deferred.

**3. Dead parameter in `_run_llm_tailoring`** (`shell/cli/_tailor.py:102`)
The `gap` parameter (typed as `object`) is accepted but never used in the function body. Either remove it or pass gap context to the tailoring prompt for better results.

**4. `complete_structured` JSON extraction fragility** (`shell/llm/client.py:112`)
LLMs frequently wrap JSON in markdown code fences (` ```json ... ``` `). A simple strip of code fences before `json.loads()` would improve reliability:
```python
raw = raw.strip()
if raw.startswith("```"):
    raw = "\n".join(raw.split("\n")[1:-1])
```

**5. `[all]` optional dep doesn't include `[bert]`** (`pyproject.toml`)
The `all` extra is `[llm,linkedin]` but `bert` is also an optional feature. Either rename to `v03` or include `bert` in `all`.

### Slop Detection ✅

- Commit messages: Clean, no AI attribution or slop markers
- Changelog: Professional, factual, no filler language
- PR description: Well-structured, accurate

### Code Quality ✅

- No code duplication detected across modules
- Type annotations consistent throughout
- Error handling follows project's exception hierarchy pattern
- Lazy imports used correctly for optional dependencies
- `__all__` exports defined in every module

### Security ✅

- API keys stripped of whitespace, not logged
- No prompt injection risk (user data in clearly delimited sections)
- HTML parsing uses `html.parser` (not `lxml`) — safe default
- `yaml.safe_load()` used for untrusted input
- File paths validated before reading

### Version Consistency ✅

| Source | Version |
|--------|---------|
| `pyproject.toml` | 0.3.0 |
| `__init__.py` | 0.3.0 |
| Runtime | 0.3.0 |
| Changelog | 0.3.0 |

**Comment by athola** [comment]
## PR Review Addendum - Deep Analysis Findings

Additional findings from parallel code review and test coverage analysis.

### Elevated to Non-Blocking (from agent reviews)

**6. Hard `bs4` import in `linkedin_fetcher.py:14`**
`from bs4 import BeautifulSoup, Tag` is a top-level import of an optional dependency. Unlike `client.py` which correctly defers `litellm` inside method bodies, `linkedin_fetcher` imports `bs4` eagerly. This is mitigated because `_import.py` lazily imports `linkedin_fetcher` inside `handle_import_command()`, so CLI startup is safe. But when `simple-resume import --linkedin` is invoked without `[linkedin]`, the error is a raw `ModuleNotFoundError` instead of a clean message. Should use the same lazy-import pattern as `client.py` or add an `is_linkedin_available()` guard.

**7. `is_current` stored as string `"true"` instead of `True`** (`linkedin_fetcher.py:113`)
`_parse_date_range` sets `entry["is_current"] = "true"` (string). The downstream `_convert_experience` in `core/importers/linkedin.py` uses `_get_str(entry, "is_current")` which happens to be truthy for any non-empty string, so it works by accident. Should be `entry["is_current"] = True`.

**8. Duplicated default model string** (`shell/llm/config.py:78` and `core/llm/protocol.py:27`)
Both `resolve_llm_config()` and `LLMConfig` define `"claude-sonnet-4-20250514"` as the default model. This creates drift risk. `resolve_llm_config` could delegate to `LLMConfig`'s field default instead.

**9. Missing `UnicodeDecodeError` handling** (`linkedin_fetcher.py:297`)
`file_path.read_text(encoding="utf-8")` for non-UTF-8 files raises `UnicodeDecodeError`, which is not caught by `handle_import_command`. It falls through to the bare `except Exception` safety net instead of producing a clean error message.

### Test Coverage Gaps (from test analysis agent)

| Priority | Gap | Severity |
|----------|-----|----------|
| 1 | `_run_llm_tailoring` success path **completely untested** | High |
| 2 | `_run_llm_tailoring` `LLMError` exception path untested | High |
| 3 | HTML experience extraction (dates, `is_current`) not asserted | Medium |
| 4 | `resolve_llm_config` with `temperature=0.0` edge case | Low |
| 5 | Short keyword filter (`len > 3`) excludes "Go", "R" etc. | Low |

The `_run_llm_tailoring` function is the primary value proposition of the tailor feature and has zero test coverage for its success path. This is the most significant test gap.

### Fragile Test Pattern

`test_tailor_no_api_key_without_report` manually pops/restores env vars instead of using `monkeypatch.delenv`. Could leak state on failure.

File: docs/plans/2026-02-24-pyok-linkedin-jobpost-design.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
+# Design: PYOK, LinkedIn Import, Job Post Tailoring
```
Comment (athola) [inline docs/plans/2026-02-24-pyok-linkedin-jobpost-design.md]: ignore `docs/plans/`, merge this information into existing documentation, and delete this from repo history
